src/python/pants/bsp/goal.py

```diff
@@ -26,15 +38,140 @@ class BSPGoal(BuiltinGoal):
+        if goal_options.server:
+            return self._run_server(
+                graph_session=graph_session,
+                union_membership=union_membership,
+            )
+        current_session_values = graph_session.scheduler_session.py_session.session_values
+        env = current_session_values[CompleteEnvironment]
+        return self._setup_bsp_connection(
+            union_membership=union_membership, env=env, options=goal_options
+        )
+
+    def _setup_bsp_connection(
+        self,
+        union_membership: UnionMembership,
+        env: Mapping[str, str],
+        options: OptionValueContainer,
     ) -> ExitCode:
+        """Setup the BSP connection file."""
+
+        build_root = BuildRoot()
+        bsp_conn_path = build_root.pathlib_path / ".bsp" / "pants.json"
+        if bsp_conn_path.exists():
+            print(
+                f"ERROR: A BSP connection file already exists at path `{bsp_conn_path}`. "
+                "Please delete that file if you intend to re-setup BSP in this repository."
+            )
+            return PANTS_FAILED_EXIT_CODE
+
+        bsp_dir = build_root.pathlib_path / ".pants.d" / "bsp"
+
+        bsp_scripts_dir = bsp_dir / "scripts"
+        bsp_scripts_dir.mkdir(exist_ok=True, parents=True)
+
+        bsp_logs_dir = bsp_dir / "logs"
+        bsp_logs_dir.mkdir(exist_ok=True, parents=True)
+
+        run_script_env_lines: list[str] = []
+        for env_var in options.runner_env_vars:
+            if "=" in env_var:
+                run_script_env_lines.append(env_var)
+            else:
+                if env_var not in env:
+                    print(
+                        f"ERROR: The `[{self.name}].runner_env_vars` option is configured to add the `{env_var}` "
+                        "environment variable to the BSP runner script using its value in  the current environment. "
+                        "That environment variable, however, is not present in the current environment. "
+                        "Please either set it in the current environment first or else configure a specific value "
+                        "in `pants.toml`."
+                    )
+                    return PANTS_FAILED_EXIT_CODE
+                run_script_env_lines.append(f"{env_var}={env[env_var]}")
+
+        run_script_env_lines_str = "\n".join(
+            [f"export {shlex.quote(line)}" for line in run_script_env_lines]
+        )
+
+        run_script_path = bsp_scripts_dir / "run-bsp.sh"
+        run_script_path.write_text(
+            textwrap.dedent(
+                f"""\
+            #!/bin/sh
+            {run_script_env_lines_str}
+            exec 2>>{shlex.quote(str(bsp_logs_dir / 'stderr.log'))}
+            env 1>&2
+            exec {shlex.quote(bin_name())} --no-pantsd {self.name} --server
+            """
+            )
+        )
+        run_script_path.chmod(0o755)
+        print(f"Wrote BSP runner script to `{run_script_path}`.")
```

Comment:
In contrast to the above errors using `print` rather than `logger.error()`, this does probably seem more appropriate to use `logger.info()`. This is all hand-wavy, but my intuition is to log because that's what we do for all other goals that say we wrote an artifact, like `package`. Whereas the errors are exception messages.

Reply:
Will do.